variable/include/scipp/variable/variable_factory.h

```diff
@@ -0,0 +1,117 @@
+    else
+      return makeVariable<T>(dims, Values(volume, core::default_init_elements));
+  }
+  DType elem_dtype(const VariableConstView &var) const override {
+    return var.dtype();
+  }
+  bool hasVariances(const VariableConstView &var) const override {
+    return var.hasVariances();
+  }
+};
+
+SCIPP_VARIABLE_EXPORT bool is_buckets(const VariableConstView &var);
+
+/// Dynamic factory for variables.
+///
+/// The factory can be used for creating variables with a dtype that is not
+/// known in the current module, e.g., dtype<bucket<Dataset>> can be used in
+/// scipp::variable. The main prupose of this is the implementation of
+/// `transform`.
+class SCIPP_VARIABLE_EXPORT VariableFactory {
+private:
+  auto bucket_dtype() const noexcept { return dtype<void>; }
+  template <class T> auto bucket_dtype(const T &var) const noexcept {
+    return is_buckets(var) ? var.dtype() : dtype<void>;
+  }
+  template <class T, class... Ts>
+  auto bucket_dtype(const T &var, const Ts &... vars) const noexcept {
+    return is_buckets(var) ? var.dtype() : bucket_dtype(vars...);
+  }
+
+public:
+  void emplace(const DType key, std::unique_ptr<AbstractVariableMaker> makes);
+  bool contains(const DType key) const noexcept;
+  bool is_buckets(const VariableConstView &var) const;
+  template <class... Parents>
+  Variable create(const DType elem_dtype, const Dimensions &dims,
+                  const bool variances, const Parents &... parents) const {
+    const auto key = bucket_dtype(parents...);
+    return m_makers.at(key == dtype<void> ? elem_dtype : key)
+        ->create(elem_dtype, dims, variances,
+                 std::vector<VariableConstView>{parents...});
+  }
+  DType elem_dtype(const VariableConstView &var) const;
+  bool hasVariances(const VariableConstView &var) const;
+  template <class T, class Var> auto values(Var &&var) const {
+    if (!is_buckets(var))
+      return var.template values<T>();
+    const auto &maker = *m_makers.at(var.dtype());
+    auto &&data = maker.data(view(var));
+    return ElementArrayView(maker.array_params(var),
+                            data.template values<T>().data());
+  }
+  template <class T, class Var> auto variances(Var &&var) const {
+    if (!is_buckets(var))
+      return var.template variances<T>();
+    const auto &maker = *m_makers.at(var.dtype());
+    auto &&data = maker.data(view(var));
+    return ElementArrayView(maker.array_params(var),
+                            data.template variances<T>().data());
+  }
+
+private:
+  VariableConstView view(const VariableConstView &var) const { return var; }
+  VariableView view(Variable &var) const { return var; }
+  VariableView view(const VariableView &var) const { return var; }
+  std::map<DType, std::unique_ptr<AbstractVariableMaker>> m_makers;
+};
+
+/// Return the global variable factory instance
+SCIPP_VARIABLE_EXPORT VariableFactory &variableFactory();
+
+} // namespace scipp::variable
```

Comment:
I don't see any tests for `VariableFactory`. Should some be added?

Reply:
Yes, still trying to make up my mind whether this should be done directly, or indirectly via `transform` (since this is just an implementation detail of transform). In fact, all the cases of "normal" dtypes (not buckets) should already be tested via this. Since we have a good number of follow-up changes that may require modification (the whole mechanism is still a bit in the flow), I would suggest to postpone addition of such tests. In particular, when we remove the support for `event_list` and the old `realign` mechanism, we will have a considerate number of tests that we can/should refactor to instead cover the bucketed data.

Reply:
ok for postponing, we should try not to forget about this though